<?xml version="1.0"?>
<sdf version="1.6">
    <world name="empty">
        <physics name="1ms" type="ignored">
            <max_step_size>0.001</max_step_size>
            <real_time_factor>1.0</real_time_factor>
        </physics>
        <plugin filename="gz-sim-physics-system" name="gz::sim::systems::Physics"></plugin>
        <plugin filename="gz-sim-user-commands-system"
            name="gz::sim::systems::UserCommands"></plugin>
        <plugin filename="gz-sim-scene-broadcaster-system"
            name="gz::sim::systems::SceneBroadcaster"></plugin>
        <plugin filename="gz-sim-contact-system" name="gz::sim::systems::Contact"></plugin>

        <light type="directional" name="sun">
            <cast_shadows>true</cast_shadows>
            <pose>0 0 10 0 0 0</pose>
            <diffuse>0.8 0.8 0.8 1</diffuse>
            <specular>0.2 0.2 0.2 1</specular>
            <attenuation>
                <range>1000</range>
                <constant>0.9</constant>
                <linear>0.01</linear>
                <quadratic>0.001</quadratic>
            </attenuation>
            <direction>-0.5 0.1 -0.9</direction>
        </light>

        <model name="ground_plane">
            <static>true</static>
            <link name="link">
                <collision name="collision">
                    <geometry>
                        <plane>
                            <normal>0 0 1</normal>
                            <size>100 100</size>
                        </plane>
                    </geometry>
                </collision>
                <visual name="visual">
                    <geometry>
                        <plane>
                            <normal>0 0 1</normal>
                            <size>100 100</size>
                        </plane>
                    </geometry>
                    <material>
                        <ambient>0.8 0.8 0.8 1</ambient>
                        <diffuse>0.8 0.8 0.8 1</diffuse>
                        <specular>0.8 0.8 0.8 1</specular>
                    </material>
                </visual>
            </link>
        </model>

        <model name="printerArm">
        <pose>0 0 0.1 0 0 0</pose>
            <link name="base_link">
                <inertial>
                    <mass>10000</mass>
                    <inertia>
                        <ixx>1e-3</ixx>
                        <ixy>0</ixy>
                        <ixz>0</ixz>
                        <iyy>1e-3</iyy>
                        <iyz>0</iyz>
                        <izz>1e-3</izz>
                    </inertia>
                </inertial>
                <pose>0 0 0 0 0 0</pose>
                <gravity>1</gravity>
                <self_collide>0</self_collide>
                <kinematic>0</kinematic>
                <enable_wind>0</enable_wind>
                <visual name='visual'>
                    <pose>0 0 0 0 0 0</pose>
                    <geometry>
                        <cylinder>
                            <radius>0.2</radius>
                            <length>0.2</length>
                        </cylinder>
                    </geometry>
                    <material>
                        <lighting>1</lighting>
                        <script>
                            <uri>file://media/materials/scripts/gazebo.material</uri>
                            <name>Gazebo/Grey</name>
                        </script>
                        <shader type='pixel'>
                            <normal_map>__default__</normal_map>
                        </shader>
                        <ambient>0.3 0.3 0.3 1</ambient>
                        <diffuse>0.7 0.7 0.7 1</diffuse>
                        <specular>0.01 0.01 0.01 1</specular>
                        <emissive>0 0 0 1</emissive>
                    </material>
                    <transparency>0</transparency>
                    <cast_shadows>1</cast_shadows>
                </visual>
                <collision name='collision'>
                    <laser_retro>0</laser_retro>
                    <max_contacts>10</max_contacts>
                    <pose>0 0 0 0 -0 0</pose>
                    <geometry>
                        <cylinder>
                            <radius>0.2</radius>
                            <length>0.2</length>
                        </cylinder>
                    </geometry>
                    <surface>
                        <friction>
                            <ode>
                                <mu>1</mu>
                                <mu2>1</mu2>
                                <fdir1>0 0 0</fdir1>
                                <slip1>0</slip1>
                                <slip2>0</slip2>
                            </ode>
                            <torsional>
                                <coefficient>1</coefficient>
                                <patch_radius>0</patch_radius>
                                <surface_radius>0</surface_radius>
                                <use_patch_radius>1</use_patch_radius>
                                <ode>
                                    <slip>0</slip>
                                </ode>
                            </torsional>
                        </friction>
                        <bounce>
                            <restitution_coefficient>0</restitution_coefficient>
                            <threshold>1e+06</threshold>
                        </bounce>
                        <contact>
                            <collide_without_contact>0</collide_without_contact>
                            <collide_without_contact_bitmask>1</collide_without_contact_bitmask>
                            <collide_bitmask>1</collide_bitmask>
                            <ode>
                                <soft_cfm>0</soft_cfm>
                                <soft_erp>0.2</soft_erp>
                                <kp>1e+13</kp>
                                <kd>1</kd>
                                <max_vel>0.01</max_vel>
                                <min_depth>0</min_depth>
                            </ode>
                            <bullet>
                                <split_impulse>1</split_impulse>
                                <split_impulse_penetration_threshold>-0.01</split_impulse_penetration_threshold>
                                <soft_cfm>0</soft_cfm>
                                <soft_erp>0.2</soft_erp>
                                <kp>1e+13</kp>
                                <kd>1</kd>
                            </bullet>
                        </contact>
                    </surface>
                </collision>
            </link>

            <link name="link1">
                <inertial>
                    <mass>10</mass>
                    <inertia>
                        <ixx>1e-3</ixx>
                        <ixy>0</ixy>
                        <ixz>0</ixz>
                        <iyy>1e-3</iyy>
                        <iyz>0</iyz>
                        <izz>1e-3</izz>
                    </inertia>
                </inertial>
                <pose relative_to="base_link">0 0 0.4 0 0 0</pose>
                <gravity>1</gravity>
                <self_collide>0</self_collide>
                <kinematic>0</kinematic>
                <enable_wind>0</enable_wind>
                <visual name='visual'>
                    <pose>0 0 0 0 0 0</pose>
                    <geometry>
                        <cylinder>
                            <radius>0.05</radius>
                            <length>0.6</length>
                        </cylinder>
                    </geometry>
                    <material>
                        <lighting>1</lighting>
                        <script>
                            <uri>file://media/materials/scripts/gazebo.material</uri>
                            <name>Gazebo/Red</name>
                        </script>
                        <shader type='pixel'>
                            <normal_map>__default__</normal_map>
                        </shader>
                        <ambient>0.3 0.3 0.3 1</ambient>
                        <diffuse>0.7 0.7 0.7 1</diffuse>
                        <specular>0.01 0.01 0.01 1</specular>
                        <emissive>0 0 0 1</emissive>
                    </material>
                    <transparency>0</transparency>
                    <cast_shadows>1</cast_shadows>
                </visual>
                <collision name='collision'>
                    <laser_retro>0</laser_retro>
                    <max_contacts>10</max_contacts>
                    <pose>0 0 0 0 -0 0</pose>
                    <geometry>
                        <cylinder>
                            <radius>0.05</radius>
                            <length>0.6</length>
                        </cylinder>
                    </geometry>
                    <surface>
                        <friction>
                            <ode>
                                <mu>1</mu>
                                <mu2>1</mu2>
                                <fdir1>0 0 0</fdir1>
                                <slip1>0</slip1>
                                <slip2>0</slip2>
                            </ode>
                            <torsional>
                                <coefficient>1</coefficient>
                                <patch_radius>0</patch_radius>
                                <surface_radius>0</surface_radius>
                                <use_patch_radius>1</use_patch_radius>
                                <ode>
                                    <slip>0</slip>
                                </ode>
                            </torsional>
                        </friction>
                        <bounce>
                            <restitution_coefficient>0</restitution_coefficient>
                            <threshold>1e+06</threshold>
                        </bounce>
                        <contact>
                            <collide_without_contact>0</collide_without_contact>
                            <collide_without_contact_bitmask>1</collide_without_contact_bitmask>
                            <collide_bitmask>1</collide_bitmask>
                            <ode>
                                <soft_cfm>0</soft_cfm>
                                <soft_erp>0.2</soft_erp>
                                <kp>1e+13</kp>
                                <kd>1</kd>
                                <max_vel>0.01</max_vel>
                                <min_depth>0</min_depth>
                            </ode>
                            <bullet>
                                <split_impulse>1</split_impulse>
                                <split_impulse_penetration_threshold>-0.01</split_impulse_penetration_threshold>
                                <soft_cfm>0</soft_cfm>
                                <soft_erp>0.2</soft_erp>
                                <kp>1e+13</kp>
                                <kd>1</kd>
                            </bullet>
                        </contact>
                    </surface>
                </collision>
            </link>

            <link name="link2">
                <inertial>
                    <mass>6</mass>
                    <inertia>
                        <ixx>1e-3</ixx>
                        <ixy>0</ixy>
                        <ixz>0</ixz>
                        <iyy>1e-3</iyy>
                        <iyz>0</iyz>
                        <izz>1e-3</izz>
                    </inertia>
                </inertial>
                <pose relative_to='link1'>0 0 0.6 0 0 0</pose>
                <gravity>1</gravity>
                <self_collide>0</self_collide>
                <kinematic>0</kinematic>
                <enable_wind>0</enable_wind>
                <visual name='visual'>
                    <pose>0 0 0 0 0 0</pose>
                    <geometry>
                        <cylinder>
                            <radius>0.025</radius>
                            <length>0.6</length>
                        </cylinder>
                    </geometry>
                    <material>
                        <lighting>1</lighting>
                        <script>
                            <uri>file://media/materials/scripts/gazebo.material</uri>
                            <name>Gazebo/Red</name>
                        </script>
                        <shader type='pixel'>
                            <normal_map>__default__</normal_map>
                        </shader>
                        <ambient>0.3 0.3 0.3 1</ambient>
                        <diffuse>0.7 0.7 0.7 1</diffuse>
                        <specular>0.01 0.01 0.01 1</specular>
                        <emissive>0 0 0 1</emissive>
                    </material>
                    <transparency>0</transparency>
                    <cast_shadows>1</cast_shadows>
                </visual>
                <collision name='collision'>
                    <laser_retro>0</laser_retro>
                    <max_contacts>10</max_contacts>
                    <pose>0 0 0 0 -0 0</pose>
                    <geometry>
                        <cylinder>
                            <radius>0.025</radius>
                            <length>0.6</length>
                        </cylinder>
                    </geometry>
                    <surface>
                        <friction>
                            <ode>
                                <mu>1</mu>
                                <mu2>1</mu2>
                                <fdir1>0 0 0</fdir1>
                                <slip1>0</slip1>
                                <slip2>0</slip2>
                            </ode>
                            <torsional>
                                <coefficient>1</coefficient>
                                <patch_radius>0</patch_radius>
                                <surface_radius>0</surface_radius>
                                <use_patch_radius>1</use_patch_radius>
                                <ode>
                                    <slip>0</slip>
                                </ode>
                            </torsional>
                        </friction>
                        <bounce>
                            <restitution_coefficient>0</restitution_coefficient>
                            <threshold>1e+06</threshold>
                        </bounce>
                        <contact>
                            <collide_without_contact>0</collide_without_contact>
                            <collide_without_contact_bitmask>1</collide_without_contact_bitmask>
                            <collide_bitmask>1</collide_bitmask>
                            <ode>
                                <soft_cfm>0</soft_cfm>
                                <soft_erp>0.2</soft_erp>
                                <kp>1e+13</kp>
                                <kd>1</kd>
                                <max_vel>0.01</max_vel>
                                <min_depth>0</min_depth>
                            </ode>
                            <bullet>
                                <split_impulse>1</split_impulse>
                                <split_impulse_penetration_threshold>-0.01</split_impulse_penetration_threshold>
                                <soft_cfm>0</soft_cfm>
                                <soft_erp>0.2</soft_erp>
                                <kp>1e+13</kp>
                                <kd>1</kd>
                            </bullet>
                        </contact>
                    </surface>
                </collision>
            </link>

            <joint name='joint1' type='revolute'>
                <parent>base_link</parent>
                <child>link1</child>
                <pose>0 0 0 0 0 0</pose>
                <axis>
                    <xyz>0 0 1</xyz>
                    <limit>
                        <lower>-1.79769e+308</lower>
                        <upper>1.79769e+308</upper>
                        <effort>-1</effort>
                        <velocity>-1</velocity>
                    </limit>
                    <dynamics>
                        <spring_reference>0</spring_reference>
                        <spring_stiffness>0</spring_stiffness>
                        <damping>0</damping>
                        <friction>0</friction>
                    </dynamics>
                </axis>
                <physics>
                    <ode>
                        <limit>
                            <cfm>0</cfm>
                            <erp>0.2</erp>
                        </limit>
                        <suspension>
                            <cfm>0</cfm>
                            <erp>0.2</erp>
                        </suspension>
                    </ode>
                </physics>
            </joint>

            <joint name='joint2' type='revolute'>
                <parent>link1</parent>
                <child>link2</child>
                <pose>0 0 0 0 0 0</pose>
                <axis>
                    <xyz>0 1 0</xyz>
                    <limit>
                        <lower>-1.79769e+308</lower>
                        <upper>1.79769e+308</upper>
                        <effort>-1</effort>
                        <velocity>-1</velocity>
                    </limit>
                    <dynamics>
                        <spring_reference>0</spring_reference>
                        <spring_stiffness>0</spring_stiffness>
                        <damping>0</damping>
                        <friction>0</friction>
                    </dynamics>
                </axis>
                <physics>
                    <ode>
                        <limit>
                            <cfm>0</cfm>
                            <erp>0.2</erp>
                        </limit>
                        <suspension>
                            <cfm>0</cfm>
                            <erp>0.2</erp>
                        </suspension>
                    </ode>
                </physics>
            </joint>

            <plugin
                filename="gz-sim-joint-trajectory-controller-system"
                name="gz::sim::systems::JointTrajectoryController">
                <joint_name>joint1</joint_name>
                <initial_position>0</initial_position>
                <position_p_gain>20</position_p_gain>
                <position_i_gain>0.4</position_i_gain>
                <position_d_gain>1.0</position_d_gain>
                <position_i_min>-1</position_i_min>
                <position_i_max>1</position_i_max>
                <position_cmd_min>-20</position_cmd_min>
                <position_cmd_max>20</position_cmd_max>

                <joint_name>joint2</joint_name>
                <initial_position>0</initial_position>
                <position_p_gain>10</position_p_gain>
                <position_i_gain>0.2</position_i_gain>
                <position_d_gain>0.5</position_d_gain>
                <position_i_min>-1</position_i_min>
                <position_i_max>1</position_i_max>
                <position_cmd_min>-10</position_cmd_min>
                <position_cmd_max>10</position_cmd_max>
            </plugin>

        </model>

    </world>
</sdf>